eval/crypt_test.go

```diff
@@ -0,0 +1,154 @@
+)
+
+type rot128 struct{}
+
+func (rot128) Encrypt(_ context.Context, plaintext []byte) ([]byte, error) {
+	for i, b := range plaintext {
+		plaintext[i] = b + 128
+	}
+	return plaintext, nil
+}
+
+func (rot128) Decrypt(_ context.Context, plaintext []byte) ([]byte, error) {
+	for i, b := range plaintext {
+		plaintext[i] = b + 128
+	}
+	return plaintext, nil
+}
+
+func TestCrypt(t *testing.T) {
+	path := filepath.Join("testdata", "crypt")
+	entries, err := os.ReadDir(path)
+	require.NoError(t, err)
+	for _, e := range entries {
+		baseName, ok := strings.CutSuffix(e.Name(), ".plaintext.yaml")
+		if !ok {
+			continue
+		}
+
+		t.Run(e.Name(), func(t *testing.T) {
+			plaintextPath := filepath.Join(path, e.Name())
+			ciphertextPath := filepath.Join(path, baseName+".ciphertext.yaml")
+
+			plaintextBytes, err := os.ReadFile(plaintextPath)
+			require.NoError(t, err)
+
+			if accept() {
+				encrypted, err := EncryptSecrets(context.Background(), baseName, plaintextBytes, rot128{})
+				require.NoError(t, err)
+
+				err = os.WriteFile(ciphertextPath, encrypted, 0o600)
+				require.NoError(t, err)
+
+				decrypted, err := DecryptSecrets(context.Background(), baseName, encrypted, rot128{})
+				require.NoError(t, err)
+
+				err = os.WriteFile(plaintextPath, decrypted, 0o600)
+				require.NoError(t, err)
+
+				return
+			}
+
+			ciphertextBytes, err := os.ReadFile(ciphertextPath)
+			require.NoError(t, err)
+
+			encrypted, err := EncryptSecrets(context.Background(), baseName, plaintextBytes, rot128{})
+			require.NoError(t, err)
+
+			require.Equal(t, ciphertextBytes, encrypted)
+
+			decrypted, err := DecryptSecrets(context.Background(), baseName, encrypted, rot128{})
+			require.NoError(t, err)
+
+			assert.Equal(t, plaintextBytes, decrypted)
+		})
+	}
+}
+
+func TestDecryptInvalidBase64(t *testing.T) {
+	const doc = `values:
+  password:
+    fn::secret:
+      ciphertext: "********"
+`
+
+	_, err := DecryptSecrets(context.Background(), "doc", []byte(doc), rot128{})
+	assert.Error(t, err)
+}
+
+type broken struct{}
+
+func (broken) Encrypt(_ context.Context, plaintext []byte) ([]byte, error) {
+	return nil, errors.New("broken")
+}
+
+func (broken) Decrypt(_ context.Context, plaintext []byte) ([]byte, error) {
+	return nil, errors.New("broken")
+}
+
+func TestEncryptBroken(t *testing.T) {
+	const doc = `values:
+  password:
+    fn::secret: hunter2
+`
+
+	_, err := EncryptSecrets(context.Background(), "doc", []byte(doc), broken{})
+	assert.Error(t, err)
+}
+
+func TestDecryptBroken(t *testing.T) {
+	const doc = `values:
+  password:
+    fn::secret:
+      ciphertext: "aHVudGVyMg=="
+`
+
+	_, err := DecryptSecrets(context.Background(), "doc", []byte(doc), broken{})
+	assert.Error(t, err)
+}
+
+func TestEncryptMalformedSecret(t *testing.T) {
+	const doc = `values:
+  password:
+    fn::secret: [ array ]
+`
+
+	_, err := EncryptSecrets(context.Background(), "doc", []byte(doc), broken{})
+	assert.NoError(t, err)
```

Comment:
Why is this NoError rather than Error? Same for the matching decrypt test.

Reply:
Ah--malformed secrets are ignored by the encrypt/decrypt paths. Error reporting will happen if/when the environment is parsed.